### DISCUSSION TITLE
How to inform Map/MapControl that a widget has changed

### DISCUSSION THREAD

**Discussion by charlenni** [Mapsui/Mapsui#2376] [Development]
If a widget or its properties change, Map or MapControl should be informed, that something changed. If this is the case, then the map has to be redrawn.

There are different methods to do this:
- Flag, that the widget is "dirty"
- INotifyPropertyChanged event
- Map is saved in the widget and Map.RefreshGraphics() could be called directly
- Any other communication tool

At the moment, we have INotifyPropertyChanged implemented.

**Comment by charlenni** [comment]
I looked into the code and would say, that it is possible to use a NeedRefresh flag for Widgets. With this, we could call from the invalidating timer e.g. Map.WidgetsNeedRefresh(), that checks all widgets, if one or more needs a refresh. In the renderers this flag is then set to false.

**Comment by charlenni** [comment]
I found, that we use this technic already for CalloutStyle with `Invalidated`. Do you think this better than the `INotifyPropertyChanged` concept?

**Comment by charlenni** [comment]
This is the follow-up to discussion in #2366.

I understand the problem with the memory leaks. Both of you suggest the dirty flag. It is the easiest one to do. We then check in the timer for redrawing, if one or more of the widgets are dirty and, if yes, redraw the map. The only point to discuss is the name of such a flag, because we should keep the naming the same in the whole project.

There are some different names that come in mind for this:
- NeedsRefresh
- NeedsRedraw
- NeedsUpdate
- UpdateRequired
- Redraw
- Dirty
- IsDirty
- Invalid
- Invalidated
- Changed
- and so on

Any of them, that is ok? Anything else?

**Reply by pauldendulk** [reply]
I saw you choose NeedsRedraw. I think that was the best option. Describes what it is doing. New developers will understand what it means.

**Comment by pauldendulk** [comment]
Here is a list of all the arguments against INotifyPropertyChanged I could think of. I might not be very balanced 😁

- INotifyPropertyChanged tends to get verbose. A single field becomes 6 lines of code.
- Ah, I know what you were thinking! We use some tool to generate them, like IL weaving or source generators. Sure, it will make the code more compact but it will make things more complex. The tools will cause bugs now and then. The build will break more often, and there will be ... things! If you have more stuff, more stuff breaks, that is a rule in life. (this still may be the best option though, we could look into how this works in Blazor).
- When introducing widgets for the very first time I mentioned that a risk is that it would develop into a UI framework. We should try to avoid rewriting Avalonia for a few overlays on our map. We should not add everything we know from those framework and write our own mini version of it. We should provide the minimal tools to make possible what we need and it should be extendable.
- Notification flows are often hard to debug, we had this problem in debugging some issues in the MapView, and I had this problem in many projects in the past.
- Notification flows can cause performance problems which are hard to track down. I call this a notification storm. One notification triggers another, or triggers notifications in a collection of items, all sending their own notifications.
- INotifyPropertyChanged allows for specific updates, just this field is changed, so update just this part of the UI. We do not use this in Mapsui. Our most important scenario is good performance while dragging and zooming. In that case we update every pixel in the map and widgets. We need do this in 16 ms. This is simple. So, we do not need the specificity of INotifyPropertyChanged, just a Refresh().
- Listeners to INotifyPropertyChanged can cause memory leaks. 
- And now my biggest fear: When we implement INotifyPropertyChanged I expect an avalanche of further feature request of things users know from their UI frameworks, like binding and for instance. This will quadruple the notification code and will make bugs even more untraceable.

**Reply by charlenni** [reply]
See PR #2422

**Comment by pauldendulk** [comment]
When thinking about this I do see one problem with the NeedsRedraw flag, and that is that it may not work when we have two MapControls, each with their own rendering. The first one will set the flag to false, and the second one will not render (not entirely sure it will work like this, but that is how it works in my mental model). Anyway. even if this is a problem I propose to keep this like this for now. It should be fixed before the final release. I want to focus on other things first.
